Clarify that partial verification is a minimum

tkfu opened the pull request:
Add some language clarifying that PV ECUs can do more verification if desired, and that this can make the system more flexible.

trishankatdatadog commented:
@tkfu Are you done? Was planning to review...

tkfu commented:
Done now--just caught a couple proofing errors.

trishankatdatadog commented on uptane-standard.md:
```diff
@@ -649,7 +649,7 @@ An Uptane-compliant ECU SHALL be able to download and verify Image metadata and
 Each ECU in a vehicle receiving over-the-air updates is either a Primary or a Secondary ECU. A Primary ECU collects and delivers vehicle manifests to the Director ({{vehicle_version_manifest}}) that contain information about which images have been installed on ECUs in the vehicle. It also verifies the time and downloads and verifies the latest metadata and images for itself and for its secondarie ...
 
-All ECUs MUST verify image metadata as specified in {{metadata_verification}} before installing an image or making it available to other ECUs. A Primary ECU MUST perform full verification ({{full_verification}}). A Secondary ECU SHOULD perform full verification if possible. If a Secondary cannot perform full verification, it SHALL perform partial verification. See the Uptane Deployment Considerat ...
+All ECUs MUST verify image metadata as specified in {{metadata_verification}} before installing an image or making it available to other ECUs. A Primary ECU MUST perform full verification ({{full_verification}}). A Secondary ECU SHOULD perform full verification if possible. If a Secondary cannot perform full verification, it SHALL perform partial verification at a minimum, but MAY perform any add ...
```
How do you "perform any additional subset of full verification"?

jhdalek55 replied:
```suggestion
All ECUs MUST verify image metadata as specified in {{metadata_verification}} before installing an image or making it available to other ECUs. A Primary ECU MUST perform full verification ({{full_verification}}). A Secondary ECU SHOULD perform full verification if possible. If a Secondary cannot perform full verification, it SHALL, at the very least, perform partial verification. In addition, it MAY also perform some steps from the full verification process. See the Uptane Deployment Considerations ({{DEPLOY}}) for a discussion of how to choose between partial and full verification.
```

jhdalek55 replied:
I don't know if this quite fixes the wording problem, but I think its a step closer.

trishankatdatadog commented on uptane-standard.md:
```diff
@@ -273,7 +273,7 @@ The following use cases provide a number of scenarios illustrating the manner in
 
 1. all primaries perform full verification;
-1. all secondaries that are updated via OTA perform full or partial verification; and
+1. all secondaries that are updated via OTA perform partial verification at a minimum; and
```
A little bit vague, because there's only minimum (partial) and maximum (full).

jhdalek55 replied:
I think this is a question of emphasis @trishankatdatadog. We want to establish partial verification as a required baseline. If you think the word "minimum" is confusing in this context, maybe it should be phrased "all secondaries that  are updated via OTA MUST at least perform partial verification" ..

jhdalek55 replied:
```suggestion
1. all secondaries that are updated via OTA MUST at least perform partial verification; and
```

jhdalek55 replied:
Here again...I think this answers the concern

trishankatdatadog commented on uptane-standard.md:
```diff
@@ -740,7 +740,7 @@ Unless the Secondary ECU has its own way of verifying the time or does not have
 #### Send metadata to Secondaries {#send_metadata_primary}
 
-The Primary SHALL send its latest downloaded metadata to all of its associated Secondaries. The metadata it sends to each Secondary MUST include all of the metadata required for verification on that Secondary. For full verification Secondaries, this includes the metadata for all four roles from both repositories, plus any delegated Targets metadata files the Secondary will recurse through to find ...
+The Primary SHALL send its latest downloaded metadata to all of its associated Secondaries. The metadata it sends to each Secondary MUST include all of the metadata required for verification on that Secondary. For full verification Secondaries, this includes the metadata for all four roles from both repositories, plus any delegated Targets metadata files the Secondary will recurse through to find ...
```
Do partial verification secondaries have a mechanism to rotate Targets key from Director? I forget. If yes, then we will need to clarify.

trishankatdatadog commented on uptane-standard.md:
```diff
@@ -818,7 +818,7 @@ The ECU SHALL create a version report as described in {{version_report}}, and se
 ### Metadata verification procedures {#metadata_verification}
 
-A Primary ECU MUST perform full verification of metadata. A Secondary ECU SHOULD perform full verification of metadata. If a Secondary cannot perform full verification, it SHALL perform partial verification instead.
+A Primary ECU MUST perform full verification of metadata. A Secondary ECU SHOULD perform full verification of metadata. If a Secondary cannot perform full verification, it SHALL perform at least partial verification instead.
```
Same reservation about "at least": there's either partial or full, I don't understand what a subset would look like.

jhdalek55 replied:
See explanation above. We are trying to establish that there is a "minimum" that must be performed. If that point is not clear, perhaps it needs to be rephrased.

jhdalek55 replied:
It's the word order in this case. See suggestion:
```suggestion
A Primary ECU MUST perform full verification of metadata. A Secondary ECU SHOULD perform full verification of metadata. If a Secondary cannot perform full verification, it SHALL, at the very least, perform partial verification.
```

trishankatdatadog commented on uptane-standard.md:
```diff
@@ -829,6 +829,12 @@ In order to perform partial verification, an ECU SHALL perform the following ste
 
+Note that this verification procedure is intended to be a minimum. A partial verification ECU MAY implement more metadata checks.
+
+For example, in many cases it is desirable to also fetch and verify Root metadata from the Director (following the procedure in {{check_root}}) before checking Targets metadata. Without this check, an ECU will not be able to verify any metadata from the Director if the Director's Targets key is rotated, possible necessitating a recall.
```
Ok, I see my question has been answered here. Needs to be linked to my question above, IMO.

jhdalek55 replied:
```suggestion
Note that this verification procedure outline the minimal number of actions that must be taken. A partial verification ECU MAY implement more metadata checks.
```

jhdalek55 replied:
I can't insert the suggestion in the right place for some reason. But, I would re-word line 832 as:

Note that this verification procedure outline the minimum number of actions that MUST be performed.

patrickvacek commented on uptane-standard.md:
```diff
@@ -829,6 +829,12 @@ In order to perform partial verification, an ECU SHALL perform the following ste
 
+Note that this verification procedure is intended to be a minimum. A partial verification ECU MAY implement more metadata checks.
+
+For example, in many cases it is desirable to also fetch and verify Root metadata from the Director (following the procedure in {{check_root}}) before checking Targets metadata. Without this check, an ECU will not be able to verify any metadata from the Director if the Director's Targets key is rotated, possible necessitating a recall.
```
```suggestion
For example, in many cases it is desirable to also fetch and verify Root metadata from the Director (following the procedure in {{check_root}}) before checking Targets metadata. Without this check, an ECU will not be able to verify any metadata from the Director if the Director's Targets key is rotated, possibly necessitating a recall.
```

jhdalek55 commented:
I think we still have some unresolved comments here. @tkfu when you get a chance can you address these.

iramcdonald commented:
Hi,

I strongly agree with @jhdalek55 that "minimum" is the *wrong* word.  I
prefer
his suggested change.

In standards specs, well-written conformance sections never use the term
"minimum".  There are simply mandatory requirements (e.g., MUST at least
perform partial verification) and recommendations (SHOULD perform full
verification).

Cheers,
- Ira


*Ira McDonald (Musician / Software Architect)Co-Chair - TCG Trusted
Mobility Solutions WG*

*Co-Chair - TCG Metadata Access Protocol SG*








*Chair - Linux Foundation Open Printing WGSecretary - IEEE-ISTO Printer
Working GroupCo-Chair - IEEE-ISTO PWG Internet Printing Protocol WGIETF
Designated Expert - IPP & Printer MIBBlue Roof Music / High North
Inchttp://sites.google.com/site/blueroofmusic
<http://sites.google.com/site/blueroofmusic>http://sites.google.com/site/highnorthinc
<http://sites.google.com/site/highnorthinc>mailto: blueroofmusic@gmail.com
<blueroofmusic@gmail.com>(permanent) PO Box 221  Grand Marais, MI 49839
906-494-2434*


On Tue, Jun 23, 2020 at 8:48 AM Lois Anne DeLong <notifications@github.com>
wrote:

> *@jhdalek55* commented on this pull request.
> ------------------------------
>
> In uptane-standard.md
> <https://github.com/uptane/uptane-standard/pull/165#discussion_r444196079>
> :
>
> > @@ -818,7 +818,7 @@ The ECU SHALL create a version report as described in {{version_report}}, and se
>
>  ### Metadata verification procedures {#metadata_verification}
>
> -A Primary ECU MUST perform full verification of metadata. A Secondary ECU SHOULD perform full verification of metadata. If a Secondary cannot perform full verification, it SHALL perform partial verification instead.
> +A Primary ECU MUST perform full verification of metadata. A Secondary ECU SHOULD perform full verification of metadata. If a Secondary cannot perform full verification, it SHALL perform at least partial verification instead.
>
> See explanation above. We are trying to establish that there is a
> "minimum" that must be performed. If that point is not clear, perhaps it
> needs to be rephrased.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/uptane/uptane-standard/pull/165#discussion_r444196079>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE33UOZX4HXXWB4PNJL2NXTRYCQCBANCNFSM4NZPJTTQ>
> .
>

trishankatdatadog commented:
Agree with @jhdalek55 and @iramcdonald, let's avoid the word "minimum" and be as clear as possible

jhdalek55 commented:
@tkfu I know you have a full plate, but is there any chance you can review the suggested changes (including a few wording changes I just proposed today, so we could close this out? Thanks.

iramcdonald commented:
Hi,

This thread is still using "minimum" - please don't.  Just say that if a
Secondary
cannot perform full verification, then it MUST perform partial
verification.  (full
stop)

Then nothing stops that Secondary from performing selective full
verification
steps, as a security enhancement.

BTW - MUST *or* SHALL - but one or the other has to be removed entirely from
the Standard and Deployment Considerations.  IETF RFCs use MUST, ISO and
ITU-T standards  use SHALL.  Take your choice please.

Cheers,
- Ira


*Ira McDonald (Musician / Software Architect)Co-Chair - TCG Trusted
Mobility Solutions WG*

*Co-Chair - TCG Metadata Access Protocol SG*








*Chair - Linux Foundation Open Printing WGSecretary - IEEE-ISTO Printer
Working GroupCo-Chair - IEEE-ISTO PWG Internet Printing Protocol WGIETF
Designated Expert - IPP & Printer MIBBlue Roof Music / High North
Inchttp://sites.google.com/site/blueroofmusic
<http://sites.google.com/site/blueroofmusic>http://sites.google.com/site/highnorthinc
<http://sites.google.com/site/highnorthinc>mailto: blueroofmusic@gmail.com
<blueroofmusic@gmail.com>(permanent) PO Box 221  Grand Marais, MI 49839
906-494-2434*


On Fri, Jul 17, 2020 at 3:58 PM Lois Anne DeLong <notifications@github.com>
wrote:

> *@jhdalek55* commented on this pull request.
> ------------------------------
>
> In uptane-standard.md
> <https://github.com/uptane/uptane-standard/pull/165#discussion_r456647052>
> :
>
> > @@ -829,6 +829,12 @@ In order to perform partial verification, an ECU SHALL perform the following ste
>  1. Load and verify the current time or the most recent securely attested time.
>  2. Download and check the Targets metadata file from the Director repository, following the procedure in {{check_targets}}.
>
> +Note that this verification procedure is intended to be a minimum. A partial verification ECU MAY implement more metadata checks.
> +
> +For example, in many cases it is desirable to also fetch and verify Root metadata from the Director (following the procedure in {{check_root}}) before checking Targets metadata. Without this check, an ECU will not be able to verify any metadata from the Director if the Director's Targets key is rotated, possible necessitating a recall.
>
> I can't insert the suggestion in the right place for some reason. But, I
> would re-word line 832 as:
>
> Note that this verification procedure outline the minimum number of
> actions that MUST be performed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/uptane/uptane-standard/pull/165#discussion_r456647052>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE33UO37F4S6TAG6Y5CCCXLR4CUPDANCNFSM4NZPJTTQ>
> .
>

jhdalek55 commented:
@iramcdonald --we will fix all the wording issues per your suggestion. I just want to be sure that is all that is needed before we can close this out.

jhdalek55 commented:
Can we try to resolve this or does it require more thought or discussion? It's been up here awhile.

patrickvacek commented on uptane-standard.md:
```diff
@@ -273,7 +273,7 @@ The following use cases provide a number of scenarios illustrating the manner in
 
 1. all primaries perform full verification;
-1. all secondaries that are updated via OTA perform partial verification at a minimum; and
+1. all secondaries that are updated via OTA MUST at least perform partial verification; and
```
Normally "Primary" and "Secondary" should be capitalized according to our style guide. I must've missed these because they are plural. Anyway, not critical and not the fault of your PR.

tkfu replied:
We can't add a MUST until 2.0, but I applied the suggestion without the MUST in it.

trishankatdatadog commented:
Let me know when this PR is ready to review, I'll help

tkfu commented:
@trishankatdatadog Ready for review.

jhdalek55 commented on uptane-standard.md:
```diff
@@ -270,10 +270,10 @@ The following use cases provide a number of scenarios illustrating the manner in
 1. all Primaries perform full verification;
-1. all Secondaries that are updated via OTA MUST at least perform partial verification; and
+1. all Secondaries that are updated via OTA at least perform partial verification; and
 1. all other ECUs that do not perform verification cannot be updated via OTA.
```
```suggestion
1. all other ECUs that do not perform any type of verification cannot be updated via OTA.
```

patrickvacek reviewed (dismissed):
Thanks!

iramcdonald commented:
Hi,

I agree with Lois's clarification.

I also note that we *can* add a SHOULD in v1.1.0 (while the MUST has to wait
for v2.0.0).

Cheers,
- Ira


*Ira McDonald (Musician / Software Architect)Co-Chair - TCG Trusted
Mobility Solutions WG*

*Co-Chair - TCG Metadata Access Protocol SG*








*Chair - Linux Foundation Open Printing WGSecretary - IEEE-ISTO Printer
Working GroupCo-Chair - IEEE-ISTO PWG Internet Printing Protocol WGIETF
Designated Expert - IPP & Printer MIBBlue Roof Music / High North
Inchttp://sites.google.com/site/blueroofmusic
<http://sites.google.com/site/blueroofmusic>http://sites.google.com/site/highnorthinc
<http://sites.google.com/site/highnorthinc>mailto: blueroofmusic@gmail.com
<blueroofmusic@gmail.com>(permanent) PO Box 221  Grand Marais, MI 49839
906-494-2434*


On Tue, Sep 1, 2020 at 11:27 AM Lois Anne DeLong <notifications@github.com>
wrote:

> *@jhdalek55* requested changes on this pull request.
>
> Looks good except for one suggested tweak
> ------------------------------
>
> In uptane-standard.md
> <https://github.com/uptane/uptane-standard/pull/165#discussion_r481228387>
> :
>
> >
>
>  1. all Primaries perform full verification;
>
> -1. all Secondaries that are updated via OTA MUST at least perform partial verification; and
>
> +1. all Secondaries that are updated via OTA at least perform partial verification; and
>
>  1. all other ECUs that do not perform verification cannot be updated via OTA.
>
>
> ⬇️ Suggested change
>
> -1. all other ECUs that do not perform verification cannot be updated via OTA.
>
> +1. all other ECUs that do not perform any type of verification cannot be updated via OTA.
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/uptane/uptane-standard/pull/165#pullrequestreview-479858851>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE33UO24OOEYFHAVPOACPRLSDUHELANCNFSM4NZPJTTQ>
> .
>

trishankatdatadog commented on uptane-standard.md:
```diff
@@ -647,9 +647,9 @@ The inventory database MAY record other information about ECUs and vehicles. It
+Each ECU receiving over-the-air updates in a vehicle is either a Primary or a Secondary ECU. A Primary ECU collects and delivers vehicle manifests to the Director ({{vehicle_version_manifest}}) that contain information about which images have been installed on ECUs in the vehicle. It also verifies the time and downloads and verifies the latest metadata and images for itself and for its Secondarie ...
 
-All ECUs MUST verify image metadata as specified in {{metadata_verification}} before installing an image or making it available to other ECUs. A Primary ECU MUST perform full verification ({{full_verification}}). A Secondary ECU SHOULD perform full verification if possible. If a Secondary cannot perform full verification, it SHALL perform partial verification. See the Uptane Deployment Considerat ...
+All ECUs MUST verify image metadata as specified in {{metadata_verification}} before installing an image or making it available to other ECUs. A Primary ECU MUST perform full verification ({{full_verification}}). A Secondary ECU SHOULD perform full verification if possible. If a Secondary cannot perform full verification, it SHALL, at the very least, perform partial verification. In addition, it  ...
```
Hmm... what "additional steps"? I don't remember. Is it in Deployment Considerations?

Seems to be related to [this](https://github.com/uptane/uptane-standard/pull/165#discussion_r437591845), but just needs tying it all up together.

tkfu replied:
I think explaining the "why" is what the deployment considerations doc is for; let's leave this for there.

There might be different reasons to choose to implement more checks; one obvious one is to check root metadata from director so you don't need a recall/out-of-band update if you need to rotate director's targets key.

jhdalek55 commented:
What is missing to complete this PR? Can we close it out this week?

tkfu commented:
> What is missing to complete this PR? Can we close it out this week?

As far as I'm concerned, nothing's missing--I believe I've addressed all comments. We just need one person to actually click the approve button, because that's the process we agreed on. @JustinCappos @mnm678 @trishankatdatadog @iramcdonald

JustinCappos reviewed (approved):
I am fine with this.  In the future, if possible, let's split out unrelated changes (primary -> Primary) into separate PRs / commits to make it cleaner.
